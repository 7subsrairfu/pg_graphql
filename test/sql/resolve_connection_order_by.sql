begin;
    create table account(
        _id serial primary key,
        id int,
        "spiritAnimal" text
    );

    insert into public.account(id, "spiritAnimal")
    values
        (1, 'bat'),
        (2, 'aardvark'),
        (3, 'aardvark'),
        (null, 'cat');


    -- Single sort

    -- AscNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: AscNullsFirst}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    -- AscNullsLast
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: AscNullsLast}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );


    -- DescNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: DescNullsFirst}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );


    -- DescNullsLast
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: DescNullsLast}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    -- Variable: AscNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "AscNullsFirst"}'
      )
    );

    -- Variable: AscNullsLast
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "AscNullsLast"}'
      )
    );

    -- Variable: DescNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "DescNullsFirst"}'
      )
    );

    -- Variable: DescNullsLast
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "DescNullsLast"}'
      )
    );

    -- Variable: Invalid
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "InvalidChoice"}'
      )
    );

    -- Variable: Missing
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               totalCount
               edges {
                 node{
                       id
                 }
               }
             }
           }
        $$,
        variables:= '{}'
      )
    );

rollback;
