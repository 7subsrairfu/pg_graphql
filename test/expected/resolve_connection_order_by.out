begin;
    create table account(
        _id serial primary key,
        id int,
        "spiritAnimal" text
    );
    insert into public.account(id, "spiritAnimal")
    values
        (1, 'bat'),
        (2, 'aardvark'),
        (3, 'aardvark'),
        (null, 'cat');
    -- Single sort
    -- AscNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: AscNullsFirst}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- AscNullsLast
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: AscNullsLast}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- DescNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: DescNullsFirst}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- DescNullsLast
    select jsonb_pretty(
        graphql.resolve($$
            {
              allAccounts(orderBy: [{id: DescNullsLast}]) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable: AscNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "AscNullsFirst"}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable: AscNullsLast
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "AscNullsLast"}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable: DescNullsFirst
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "DescNullsFirst"}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable: DescNullsLast
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "DescNullsLast"}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable: Invalid
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": "InvalidChoice"}'
      )
    );
                jsonb_pretty                
--------------------------------------------
 {                                         +
     "data": null,                         +
     "errors": [                           +
         "Invalid order by clause variable"+
     ]                                     +
 }
(1 row)

    -- Variable: Missing
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: OrderByDirection)
           {
             allAccounts(orderBy: [{id: $direction}]) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{}'
      )
    );
                jsonb_pretty                
--------------------------------------------
 {                                         +
     "data": null,                         +
     "errors": [                           +
         "Invalid order by clause variable"+
     ]                                     +
 }
(1 row)

    -- Variable Whole Param: AscNullsFirst 
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": [{"id": "AscNullsFirst"}]}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable Whole Param: AscNullsLast
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": [{"id": "AscNullsLast"}]}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable Whole Param: DescNullsFirst 
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": [{"id": "DescNullsFirst"}]}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable Whole Param: DescullsLast 
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": [{"id": "DescNullsLast"}]}'
      )
    );
            jsonb_pretty            
------------------------------------
 {                                 +
     "data": {                     +
         "allAccounts": {          +
             "edges": [            +
                 {                 +
                     "node": {     +
                         "id": 3   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 2   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": 1   +
                     }             +
                 },                +
                 {                 +
                     "node": {     +
                         "id": null+
                     }             +
                 }                 +
             ]                     +
         }                         +
     },                            +
     "errors": [                   +
     ]                             +
 }
(1 row)

    -- Variable Whole Param: Invalid, null
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": null}'
      )
    );
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "data": null,                            +
     "errors": [                              +
         "Invalid value for ordering variable"+
     ]                                        +
 }
(1 row)

    -- Variable Whole Param: Invalid, empty list
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsOrdered($direction: [AccountOrderBy])
           {
             allAccounts(orderBy: $direction) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"direction": []}'
      )
    );
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "data": null,                            +
     "errors": [                              +
         "Invalid value for ordering variable"+
     ]                                        +
 }
(1 row)

rollback;
